File: solutions/ruby/locomotive-engineer/1/locomotive_engineer.rb
class LocomotiveEngineer
  def self.generate_list_of_wagons(*args)
    args
  end

  def self.fix_list_of_wagons(each_wagons_id, missing_wagons)
    # Step 1: Move the first two items to the end
    reordered = each_wagons_id[2..] + each_wagons_id[0, 2]
  
    # Step 2: Insert missing wagons after the locomotive (1)
    loco_index = reordered.index(1)
    reordered.insert(loco_index + 1, *missing_wagons)
  
    reordered
  end


  def self.add_missing_stops(route, **stops)
    # Extract stop names in order of their keys
    ordered_stops = stops.sort_by { |key, _| key.to_s }.map { |_, city| city }
  
    # Return updated route hash
    route.merge(stops: ordered_stops)
  end


  def self.extend_route_information(route, more_route_information)
    route.merge(more_route_information)
  end
end

File: solutions/python/locomotive-engineer/1/locomotive_engineer.py
"""Functions which helps the locomotive engineer to keep track of the train."""


def get_list_of_wagons(*args) -> list:
    """Return a list of wagons.
    """
    return list(args)


def fix_list_of_wagons(each_wagons_id, missing_wagons):
    """Fix the list of wagons.
    """
    locomotive_id = 1
    locomotive_index = each_wagons_id.index(locomotive_id)
    left, right = each_wagons_id[:locomotive_index], each_wagons_id[locomotive_index+1: ]
    return [locomotive_id, *missing_wagons, *right, *left]

def add_missing_stops(routes, **kwargs):
    """Add missing stops to route dict.
    """
    routes['stops'] = list(kwargs.values())
    return routes


def extend_route_information(route, more_route_information):
    """Extend route information with more_route_information.
    """
    return {**route, **more_route_information}


def fix_wagon_depot(wagons_rows):
    """Fix the list of rows of wagons.
    """
    return [list(row) for row in zip(*wagons_rows)]
